File: fixtures/laminaria-semantic-substrate-prototype/nim-src/add_or_double.nim
## Semantic contract, see ../CONTRACT.md. Real Nim implementation of the
## workload the candidate representation (../substrate/) is cross-checked
## against. Prints "a,b,use_double,result" for the same fixed set of test
## inputs as ../rust-src/add_or_double.rs, so ../trace.sh can diff them
## byte-for-byte.

proc double(x: int32): int32 =
  x +% x

proc addOrDouble(a, b, useDouble: int32): int32 =
  if useDouble != 0'i32:
    double(a)
  else:
    a +% b

let testInputs = [
  (3'i32, 4'i32, 0'i32),
  (3'i32, 4'i32, 1'i32),
  (int32.high, 1'i32, 0'i32),
  (-5'i32, 10'i32, 1'i32),
]

for (a, b, useDouble) in testInputs:
  echo a, ",", b, ",", useDouble, ",", addOrDouble(a, b, useDouble)

File: fixtures/laminaria-semantic-substrate-prototype/rust-src/add_or_double.rs
//! Semantic contract, see ../CONTRACT.md. Real Rust implementation of the
//! workload the candidate representation (../substrate/) is cross-checked
//! against. Prints "a,b,use_double,result" for a fixed set of test inputs,
//! matching the format ../substrate's reference evaluator and LLVM-IR
//! projection also print, so ../trace.sh can diff them byte-for-byte.

fn double(x: i32) -> i32 {
    x.wrapping_add(x)
}

fn add_or_double(a: i32, b: i32, use_double: i32) -> i32 {
    if use_double != 0 {
        double(a)
    } else {
        a.wrapping_add(b)
    }
}

const TEST_INPUTS: &[(i32, i32, i32)] = &[
    (3, 4, 0),
    (3, 4, 1),
    (i32::MAX, 1, 0),
    (-5, 10, 1),
];

fn main() {
    for &(a, b, use_double) in TEST_INPUTS {
        println!("{a},{b},{use_double},{}", add_or_double(a, b, use_double));
    }
}
